railgun/manager/manager.go

```diff
@@ -0,0 +1,205 @@
+	"github.com/kong/kubernetes-ingress-controller/railgun/apis/configuration/v1alpha1"
+	configurationv1alpha1 "github.com/kong/kubernetes-ingress-controller/railgun/apis/configuration/v1alpha1"
+	configurationv1beta1 "github.com/kong/kubernetes-ingress-controller/railgun/apis/configuration/v1beta1"
+	"github.com/kong/kubernetes-ingress-controller/railgun/controllers"
+	kongctrl "github.com/kong/kubernetes-ingress-controller/railgun/controllers/configuration"
+	"k8s.io/apimachinery/pkg/runtime"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	clientgoscheme "k8s.io/client-go/kubernetes/scheme"
+	"k8s.io/client-go/tools/clientcmd"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/healthz"
+	"sigs.k8s.io/controller-runtime/pkg/log/zap"
+)
+
+// Config collects all configuration that the controller manager takes from the environment.
+// BUG: the above is not 100% accurate today - controllers read some settings from environment variables directly
+type Config struct {
+	MetricsAddr          string
+	EnableLeaderElection bool
+	ProbeAddr            string
+	KongURL              string
+	FilterTag            string
+	Concurrency          int
+	SecretName           string
+	SecretNamespace      string
+	KubeconfigPath       string
+
+	ZapOptions zap.Options
+}
+
+// RegisterFlags binds the provided Config to commandline flags.
+func RegisterFlags(c *Config, flagSet *flag.FlagSet) {
+	flagSet.StringVar(&c.MetricsAddr, "metrics-bind-address", ":8080", "The address the metric endpoint binds to.")
+	flagSet.StringVar(&c.ProbeAddr, "health-probe-bind-address", ":8081", "The address the probe endpoint binds to.")
+	flagSet.BoolVar(&c.EnableLeaderElection, "leader-elect", false,
+		"Enable leader election for controller manager. "+
+			"Enabling this will ensure there is only one active controller manager.")
+	flagSet.StringVar(&c.KongURL, "kong-url", "http://localhost:8001", "TODO")
+	flagSet.StringVar(&c.FilterTag, "kong-filter-tag", "managed-by-railgun", "TODO")
+	flagSet.IntVar(&c.Concurrency, "kong-concurrency", 10, "TODO")
+	flagSet.StringVar(&c.SecretName, "secret-name", "kong-config", "TODO")
+	flagSet.StringVar(&c.SecretNamespace, "secret-namespace", controllers.DefaultNamespace, "TODO")
+	flagSet.StringVar(&c.KubeconfigPath, "kubeconfig", "", "Path to the kubeconfig file.")
+
+	c.ZapOptions.BindFlags(flagSet)
+}
+
+// Run starts the controller manager and blocks until it exits.
+func Run(ctx context.Context, c *Config) error {
+	ctrl.SetLogger(zap.New(zap.UseFlagOptions(&c.ZapOptions)))
+	setupLog := ctrl.Log.WithName("setup")
+
+	scheme := runtime.NewScheme()
+	utilruntime.Must(clientgoscheme.AddToScheme(scheme))
+	utilruntime.Must(konghqcomv1.AddToScheme(scheme))
+	utilruntime.Must(configurationv1alpha1.AddToScheme(scheme))
+	utilruntime.Must(configurationv1beta1.AddToScheme(scheme))
+
+	// TODO: we might want to change how this works in the future, rather than just assuming the default ns
```

Comment:
just a note, we have a follow up for the TODOs in here: https://github.com/Kong/kubernetes-ingress-controller/issues/1114

Reply:
This is moved verbatim from `main()`

Reply:
no action needed, just a note on the PR :+1: